File: source-connector/src/main/java/sourceconnector/repository/offset/InternalOffsetRecordRepository.java
package sourceconnector.repository.offset;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import offsetmanager.domain.file.FileKey;
import offsetmanager.domain.file.factory.FileKeyParser;
import offsetmanager.domain.offset.DefaultOffsetRecord;
import offsetmanager.domain.offset.OffsetRecord;
import org.apache.kafka.clients.admin.AdminClient;
import org.apache.kafka.clients.admin.DescribeTopicsResult;
import org.apache.kafka.clients.admin.TopicDescription;
import org.apache.kafka.clients.consumer.Consumer;
import org.apache.kafka.clients.consumer.ConsumerRecord;
import org.apache.kafka.common.KafkaFuture;
import org.apache.kafka.common.TopicPartition;
import org.apache.kafka.common.utils.Utils;
import sourceconnector.service.offset.OffsetRecordRepository;

import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.util.*;
import java.util.concurrent.ExecutionException;
import java.util.stream.Collectors;

@Slf4j
@RequiredArgsConstructor
public class InternalOffsetRecordRepository implements OffsetRecordRepository {
  private final Consumer<String, Long> consumer;
  private final AdminClient adminClient;
  private final String offsetTopic;
  private final Duration timeout = Duration.ofMillis(100);

  @Override
  public Optional<OffsetRecord> findLastOffsetRecord(FileKey fileKey) {
    int partition = this.getPartitionsForTopic(fileKey);
    TopicPartition topicPartition = new TopicPartition(offsetTopic, partition);
    this.consumer.assign(List.of(topicPartition));
    long currentOffset = this.consumer.beginningOffsets(List.of(topicPartition)).get(topicPartition);
    long endOffset = this.consumer.endOffsets(List.of(topicPartition)).get(topicPartition);

    Optional<OffsetRecord> lastOffsetRecord = Optional.empty();

    while (currentOffset < endOffset) {
      this.consumer.seek(topicPartition, currentOffset);

      List<ConsumerRecord<String, Long>> recordList = this.consumer
        .poll(timeout)
        .records(topicPartition);

      if (recordList.isEmpty()) break; // Should not empty

      long lastOffset = recordList
        .stream()
        .max(Comparator.comparingLong(ConsumerRecord::offset))
        .map(ConsumerRecord::offset)
        .get();

      currentOffset = lastOffset + 1;

      Optional<OffsetRecord> currentLastOffsetRecord = recordList
        .stream()
        .filter(record -> record.key().equals(fileKey.get()))
        .max(Comparator.comparingLong(ConsumerRecord::offset))
        .map(record -> new DefaultOffsetRecord(
          FileKeyParser.parse(record.key()),
          record.value()
        ));

      if (currentLastOffsetRecord.isPresent()) {
        lastOffsetRecord = currentLastOffsetRecord;
      }
    }

    return lastOffsetRecord;
  }

  @Override
  public List<OffsetRecord> findLastOffsetRecords(List<FileKey> keys) {
    if (keys.isEmpty()) return Collections.emptyList();

    Map<Integer, List<FileKey>> keysByPartition = keys.stream()
      .collect(Collectors.groupingBy(this::getPartitionsForTopic));

    Map<FileKey, OffsetRecord> keyOffsetMap = new HashMap<>();
    // Iterate through each partition
    for (Map.Entry<Integer, List<FileKey>> entry: keysByPartition.entrySet()) {
      int partition = entry.getKey();
      Set<FileKey> fileKeySet = new HashSet<>(entry.getValue());

      TopicPartition topicPartition = new TopicPartition(offsetTopic, partition);
      this.consumer.assign(List.of(topicPartition));

      long currentOffset = this.consumer.beginningOffsets(List.of(topicPartition)).get(topicPartition);
      long endOffset = this.consumer.endOffsets(List.of(topicPartition)).get(topicPartition);

      while (currentOffset < endOffset) {
        consumer.seek(topicPartition, currentOffset);
        List<ConsumerRecord<String, Long>> recordList = this.consumer
          .poll(timeout)
          .records(topicPartition);

        if (recordList.isEmpty()) break;

        currentOffset = recordList.getLast().offset() + 1;

        for (var record: recordList) {
          FileKey fileKey = FileKeyParser.parse(record.key());
          if (!fileKeySet.contains(fileKey)) continue;
          long offset = record.value();
          // record last offset for each fileKey
          keyOffsetMap.put(fileKey, new DefaultOffsetRecord(fileKey, offset));
        }
      }

    }

    return new ArrayList<>(keyOffsetMap.values());
  }

  private int getPartitionsForTopic(FileKey fileKey){
    // get partition count of topic
    DescribeTopicsResult result = adminClient.describeTopics(Collections.singletonList(offsetTopic));
    Map<String, KafkaFuture<TopicDescription>> futures = result.topicNameValues();
    try {
      TopicDescription description = futures.get(offsetTopic).get();
      int numPartitions = description.partitions().size();
      byte[] serializedKey = fileKey.get().getBytes(StandardCharsets.UTF_8);
      return Utils.toPositive(Utils.murmur2(serializedKey)) % numPartitions;
    } catch (ExecutionException | InterruptedException e) {
      log.error("Failed to get partitions for topic {}", offsetTopic, e);
      throw new RuntimeException(e.getMessage());
    }
  }

  @Override
  public void close() throws Exception {
    this.consumer.close();
    this.adminClient.close();
  }
}

File: source-connector/src/main/kotlin/sourceconnector/repository/offset/InternalOffsetRecordRepository.kt
package sourceconnector.repository.offset

import offsetmanager.domain.file.FileKey
import offsetmanager.domain.file.factory.FileKeyParser
import offsetmanager.domain.offset.DefaultOffsetRecord
import offsetmanager.domain.offset.OffsetRecord
import org.apache.kafka.clients.admin.AdminClient
import org.apache.kafka.clients.consumer.Consumer
import org.apache.kafka.clients.consumer.ConsumerRecord
import org.apache.kafka.common.TopicPartition
import org.apache.kafka.common.utils.Utils
import org.slf4j.LoggerFactory
import sourceconnector.service.offset.OffsetRecordRepository
import java.nio.charset.StandardCharsets
import java.time.Duration
import java.util.concurrent.ExecutionException

class InternalOffsetRecordRepository(
  private val consumer: Consumer<String, Long>,
  private val adminClient: AdminClient,
  private val offsetTopic: String,
  private val timeout: Duration = Duration.ofMillis(100)
) : OffsetRecordRepository {

  private val log = LoggerFactory.getLogger(InternalOffsetRecordRepository::class.java)

  override fun findLastOffsetRecord(key: FileKey): OffsetRecord? {
    val partition = this.getPartitionsForTopic(key)
    val topicPartition = TopicPartition(offsetTopic, partition)
    this.consumer.assign(listOf(topicPartition))
    val beginOffset: Long = this.consumer.beginningOffsets(listOf(topicPartition))[topicPartition]!!
    val endOffset: Long = this.consumer.endOffsets(listOf(topicPartition))[topicPartition]!!

    if (beginOffset >= endOffset) {
      return null
    }

    consumer.seek(topicPartition, beginOffset)
    var lastOffsetRecord: OffsetRecord? = null

    var lastConsumedOffset: Long = Long.MIN_VALUE
    while (lastConsumedOffset < endOffset) {

      val recordList = this.consumer
        .poll(timeout)
        .records(topicPartition)

      lastOffsetRecord = recordList
        .filter { it.key() == key.get() }
        .maxBy { record: ConsumerRecord<String, Long> -> record.offset() }
        .let {
          DefaultOffsetRecord(
            FileKeyParser.parse(it.key()),
            it.value()
          )
        }

      lastConsumedOffset = consumer.position(topicPartition)
    }

    return lastOffsetRecord
  }

  override fun findLastOffsetRecords(keys: List<FileKey>): List<OffsetRecord> {
    if (keys.isEmpty()) return  emptyList()

    val keysByPartition: Map<Int, List<FileKey>> = keys.groupBy { fileKey -> getPartitionsForTopic(fileKey) }

    val keyOffsetMap: MutableMap<FileKey, OffsetRecord> = mutableMapOf()
    // Iterate through each partition
    for ((partition, fileKeys) in keysByPartition.entries) {
      val fileKeySet: Set<FileKey> = fileKeys.toSet()

      val topicPartition = TopicPartition(offsetTopic, partition)
      this.consumer.assign(listOf(topicPartition))
      val beginOffset: Long = this.consumer.beginningOffsets(listOf(topicPartition))[topicPartition]!!
      val endOffset: Long = this.consumer.endOffsets(listOf(topicPartition))[topicPartition]!!

      if (beginOffset >= endOffset) continue

      consumer.seek(topicPartition, beginOffset)

      var lastConsumedOffset: Long = Long.MIN_VALUE
      while (lastConsumedOffset < endOffset) {
        val recordList: List<ConsumerRecord<String, Long>> = this.consumer
          .poll(timeout)
          .records(topicPartition)

        for (record in recordList) {
          val fileKey = FileKeyParser.parse(record.key())
          if (!fileKeySet.contains(fileKey)) continue
          val offset: Long = record.value()!!
          keyOffsetMap[fileKey] = DefaultOffsetRecord(fileKey, offset)
        }
        lastConsumedOffset = consumer.position(topicPartition)
      }
    }

    return keyOffsetMap.values.toList()
  }

  private fun getPartitionsForTopic(fileKey: FileKey): Int {
    // get partition count of topic
    val result = adminClient.describeTopics(listOf(offsetTopic))
    val futures = result.topicNameValues()
    try {
      val description = futures[offsetTopic]!!.get()
      val numPartitions = description.partitions().size
      val serializedKey = fileKey.get().toByteArray(StandardCharsets.UTF_8)
      return Utils.toPositive(Utils.murmur2(serializedKey)) % numPartitions
    } catch (e: ExecutionException) {
      log.error("Failed to get partitions for topic {}", offsetTopic, e)
      throw RuntimeException(e.message)
    } catch (e: InterruptedException) {
      log.error("Failed to get partitions for topic {}", offsetTopic, e)
      throw RuntimeException(e.message)
    }
  }

  @Throws(Exception::class)
  override fun close() {
    this.consumer.close()
    this.adminClient.close()
  }
}
